core/src/subsystems/flywheel.cpp

```diff
@@ -17,56 +19,129 @@
+  wheel->setPIDTarget(wheel->getRPM());  // get the pid from the wheel and set its target to the RPM stored in the wheel.
   while(true) {
-    wheel->getPID()->update(wheel->getMotors()->velocity(velocityUnits::rpm));        // check the current velocity and update the PID with it.
-    wheel->getMotors()->spin(fwd, wheel->getPID()->get() * 12, voltageUnits::volt);   // set the motors to whatever PID tells them to do
+    wheel->updatePID(wheel->getVelocity_RPM());        // check the current velocity and update the PID with it.
+    wheel->spin_raw(wheel->getPIDValue(), fwd);   // set the motors to whatever PID tells them to do
     vexDelay(20);
   }
   return 0; // only here to make the compiler SHUT UP
 }
 
-// starts / restarts RPM thread at new value
-void Flywheel::spinRPM(int inputRPM) {
-  if(inputRPM != RPM) {
-    RPM = inputRPM;
-    rpmTask.stop();
-    rpmTask = task(spinRPMThread, this);
+//  Runs a BANG BANG variant. I have NO IDEA if it will work properly but I have NOTHING to test this on.
+int spinRPMTask_BangBang(void* wheelPointer) {
+  Flywheel* wheel = (Flywheel*) wheelPointer; 
+  while(true) {
+    // if it below the RPM, go, otherwise don't
+    if(wheel->getVelocity_RPM() < wheel->getRPM()) { 
+      wheel->spin_raw(1, fwd);
+    }   
+    else { wheel->stopMotors(); }
+    vexDelay(20);
   }
+  return 0; // only here to make the compiler SHUT UP
 }
 
-// return the current value that the RPM should be set to;
-double Flywheel::getRPM() { return RPM; }
+// Runs a Feedforward variant; FUNCTION STUB
+int spinRPMTask_Feedforward(void* wheelPointer) { return 0; }
+
+// Runs a PID + Feedforward variant; FUNCTION STUB
+int spinRPMTask_PID_Feedforward(void* wheelPointer) { return 0; }
+
+// Runs a Take Back Half variant; FUNCTION STUB
+int spinRPMTask_TBH(void* wheelPointer) { return 0; }
+
+// Runs a 'Moving average filter with above closed loop systems' variant, whatever that means; FUNCTION STUB
+int spinRPMTask_ClosedLoop(void* wheelPointer) { return 0; }
+
+
+/*********************************************************
+*         SPINNERS AND STOPPERS
+*********************************************************/
 
-// Returns a POINTER TO the PID; only for use in the spinRPMThread function.
-PID* Flywheel::getPID() { return &pid; }
+/* Spin motors using voltage; defaults forward at 12 volts
+* @param speed - speed (between -1 and 1) to set the motor
+* @param dir - direction that the motor moves in; defaults to forward
+*/
+void Flywheel::spin_raw(double speed, directionType dir){
+  motors.spin(dir, speed * 12, voltageUnits::volt);
+}
 
-// Returns a POINTER TO the motors; only for use in the spinRPMThread function.
-motor_group* Flywheel::getMotors() { return &motors; }
+/* starts / restarts RPM thread at new value
+* @param inputRPM - set the current RPM
+*/
+void Flywheel::spinRPM(int inputRPM) {
+  // only run if the RPM is different or it isn't already running
+  if(inputRPM != RPM || !taskRunning) {
```

Comment:
While this would probably work for instant changes in RPM setpoint, if someone wanted to ramp to a value the task scheduler would probably have a stroke creating and destroying the task every  time it loops. See above for my recommendation about creating the task once in the constructor

Reply:
One thing I like to do is design it as if it wasn't an asynchronous class, and you had some sort of `update()` method that did the calculations and set the motors that'd the programmer would call in the main loop. Then for the task, all you'd do is have that `update()` function with a delay

Reply:
Fixed (probably); Dana says it _looks_ like it will work.